File: src/lib/triage/classifier.ts
/**
 * Three-layer triage classifier orchestrator.
 * Adapted from src/lib/triage/classifier.js — accepts supabase client as parameter.
 * Pipeline: Layer 1 (regex) → Layer 2 (LLM, if ambiguous) → Layer 3 (owner rules).
 */

import type { SupabaseClient } from '@supabase/supabase-js';
import { runKeywordClassifier } from './layer1-keywords.js';
import { runLLMScorer } from './layer2-llm.js';
import { applyOwnerRules } from './layer3-rules.js';

export async function classifyCall(
  supabase: SupabaseClient,
  {
    transcript,
    tenant_id,
    detected_service = null,
  }: {
    transcript: string;
    tenant_id: string;
    detected_service?: string | null;
  },
): Promise<{ urgency: string; confidence: string; layer: string; reason?: string }> {
  if (!transcript || transcript.length < 10) {
    return { urgency: 'routine', confidence: 'low', layer: 'layer1' };
  }

  const layer1Result = runKeywordClassifier(transcript);

  if (layer1Result.confident) {
    const layer3Result = await applyOwnerRules(supabase, layer1Result.result, tenant_id, detected_service);
    const finalLayer = layer3Result.escalated ? 'layer3' : 'layer1';
    return { urgency: layer3Result.urgency, confidence: 'high', layer: finalLayer };
  }

  const layer2Result = await runLLMScorer(transcript);
  const layer3Result = await applyOwnerRules(supabase, layer2Result.urgency, tenant_id, detected_service);
  const finalLayer = layer3Result.escalated ? 'layer3' : 'layer2';

  return {
    urgency: layer3Result.urgency,
    confidence: layer2Result.confidence ?? 'low',
    layer: finalLayer,
    reason: layer2Result.reason,
  };
}

File: src/lib/triage/classifier.py
from supabase import Client

from .layer1_keywords import run_keyword_classifier
from .layer2_llm import run_llm_scorer
from .layer3_rules import apply_owner_rules

VALID_URGENCIES = {"emergency", "routine", "high_ticket"}


def _sanitize_urgency(urgency: str) -> str:
    """Ensure urgency is a valid DB enum value."""
    if urgency in VALID_URGENCIES:
        return urgency
    return "routine"


async def classify_call(
    supabase: Client,
    *,
    transcript: str | None,
    tenant_id: str,
    detected_service: str | None = None,
) -> dict:
    if not transcript or len(transcript) < 10:
        return {"urgency": "routine", "confidence": "low", "layer": "layer1"}

    layer1_result = run_keyword_classifier(transcript)

    if layer1_result["confident"]:
        layer3_result = await apply_owner_rules(
            supabase, layer1_result["result"], tenant_id, detected_service
        )
        final_layer = "layer3" if layer3_result["escalated"] else "layer1"
        return {
            "urgency": layer3_result["urgency"],
            "confidence": "high",
            "layer": final_layer,
        }

    layer2_result = await run_llm_scorer(transcript)
    layer2_urgency = _sanitize_urgency(layer2_result.get("urgency", "routine"))
    layer3_result = await apply_owner_rules(
        supabase, layer2_urgency, tenant_id, detected_service
    )
    final_layer = "layer3" if layer3_result["escalated"] else "layer2"

    return {
        "urgency": _sanitize_urgency(layer3_result["urgency"]),
        "confidence": layer2_result.get("confidence", "low"),
        "layer": final_layer,
        "reason": layer2_result.get("reason"),
    }
